### PR TITLE
Journaling

### PR DESCRIPTION
**Pull request by antoniolm23**
Added journaling configurability plus ability to run even without journaling system

### CODE REVIEW

File: PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java [hunk 1]
```diff
@@ -81,11 +78,11 @@
     private static Logger log = Logger.getLogger( PolicyDecisionPoint.class.getName() );
 
     private PDPConfig pdpConfig;
-    private Journal journal = null;
+    private Optional<JournalingInterface> journalInterface;
 
     public PolicyDecisionPoint( PdpProperties properties ) {
         super( properties );
-        buildJournal( properties.getJournalDir() );
+        journalInterface = it.cnr.iit.ucs.journaling.JournalBuilder.build( properties );
```
Comment (tezcan10) [inline PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java]: why do you need to declare it with full package definition here

File: PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java [hunk 2]
```diff
@@ -286,13 +280,13 @@ private ResponseCtx getResponseCtx( int result, String code, String message ) {
...
-        try {
-            File file = new File( journalDir ); // TODO UCS-33 NOSONAR
-            FileUtility.createPathIfNotExists( file );
-            journal = JournalBuilder.of( file ).open();
-        } catch( Exception e ) {
-            throw new IllegalArgumentException( e.getMessage() );
+    private void write( String policy, String request, String response ) {
+        if( !journalInterface.isPresent() ) {
+            return;
+        } else {
```
Comment (tezcan10) [inline PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java]: you do not really need an else here if you do an early return when not present

File: UCSRest/src/test/resources/application.properties [hunk 3]
```diff
@@ -1,7 +1,8 @@
 server.port=9998
 
 ucs.core.uri=http://localhost:9998
-ucs.core.journal-dir=/tmp/ucf
+ucs.core.journal-path=/tmp/ucf
```
Comment (tezcan10) [inline UCSRest/src/test/resources/application.properties]: funny characters at the end of each line in this file

File: UCSLibraries/src/main/java/it/cnr/iit/utility/errorhandling/Reject.java [hunk 4]
```diff
@@ -270,7 +270,7 @@ public static void ifNull( Object... objects ) {
         }
     }
 
-    public static void ifNull( String... objects ) {
+    public static void ifNullStringArray( String... objects ) {
```
Comment (tezcan10) [inline UCSLibraries/src/main/java/it/cnr/iit/utility/errorhandling/Reject.java]: why is this necessary? I mean a method for type String, isnt the existing one recognise the String as object

File: PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java (hunk 1 above)
Comment (antoniolm23) [reply PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java]: yes, I noticed it imediately after the pull request, if you update you should not see it anymore

File: PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java (hunk 2 above)
Comment (antoniolm23) [reply PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java]: as before, I've noticed it after the pull request was done

File: UCSLibraries/src/main/java/it/cnr/iit/utility/errorhandling/Reject.java (hunk 4 above)
Comment (antoniolm23) [reply UCSLibraries/src/main/java/it/cnr/iit/utility/errorhandling/Reject.java]: yes this can be removed

File: UCSRest/src/test/resources/application.properties (hunk 3 above)
Comment (hexec) [reply UCSRest/src/test/resources/application.properties]: yeah this is a carriage-return character probably generated by windows. use vim and then :s/^M$//

File: PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java [hunk 5]
```diff
@@ -55,7 +58,7 @@
 public final class PIPReader extends PIPBase {
 
     private static Logger log = Logger.getLogger( PIPReader.class.getName() );
-    private PIPJournalHelper journal;
+    private Optional<JournalingInterface> journal;
```
Comment (hexec) [inline PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java]: we should use optional only in initialization or hide the fact that the journal is disabled inside the journal

File: PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java [hunk 6]
```diff
@@ -271,6 +273,23 @@ private final void setFilePath( String filePath ) {
         Reject.ifBlank( this.filePath );
     }
 
+    private void logOperation( String... strings ) {
```
Comment (hexec) [inline PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java]: maybe we can hide these functions inside the journal?
logOperation
format
each component implementing a journal would have to replicate these functions

File: PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java [hunk 7]
```diff
@@ -271,6 +273,23 @@ private final void setFilePath( String filePath ) {
...
     }
 
+    private void logOperation( String... strings ) {
+        if( !journal.isPresent() ) {
+            return;
+        }
+        journal.get().logString( format( strings ) );
+    }
+
+    private String format( String... strings ) {
```
Comment (hexec) [inline PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java]: maybe we can use a more descriptive name : format -> formatJournalMessage

File: PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java (hunk 6 above)
Comment (hexec) [inline PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java]: Every JournalingInterface object could wrap a Optional<Journal> inside and if it is present it will be used.
In this way we won't need to replicate this function in any component that needs a journal.
if no journal is configured in the component it can be fatal or not but it is decided in the initialization of the actual component.

File: PIPReader/src/test/java/it/cnr/iit/ucs/properties/UCFPipProperties.java [hunk 8]
```diff
@@ -61,4 +58,19 @@ public void setAttributes( List<Map<String, String>> attributes ) {
         return null;
     }
 
+    @Override
+    public String getJournalPath() {
+        return journalDir;
+    }
+
+    @Override
+    public String getJournalProtocol() {
```
Comment (hexec) [inline PIPReader/src/test/java/it/cnr/iit/ucs/properties/UCFPipProperties.java]: we could put this in the base property class.
each component could have a journal configured

File: PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java [hunk 9]
```diff
@@ -105,10 +103,7 @@ public PDPEvaluation evaluate( RequestWrapper request, PolicyWrapper policy ) {
         try {
             PolicyFinder policyFinder = getPolicyFinder( policy );
             ResponseCtx responseCtx = evaluate( request.getRequest(), policyFinder );
-            journal.write( policy.getPolicy().getBytes(), WriteType.ASYNC );
-            journal.write( request.getRequest().getBytes(), WriteType.ASYNC );
-            journal.write( responseCtx.encode().getBytes(), WriteType.ASYNC );
-            journal.sync();
+            write( policy.getPolicy(), request.getRequest(), responseCtx.encode() );
```
Comment (hexec) [inline PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java]: very nice :D

File: PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java [hunk 10]
```diff
@@ -286,13 +281,11 @@ private ResponseCtx getResponseCtx( int result, String code, String message ) {
...
     }
 
-    private void buildJournal( String journalDir ) {
-        try {
-            File file = new File( journalDir ); // TODO UCS-33 NOSONAR
-            FileUtility.createPathIfNotExists( file );
-            journal = JournalBuilder.of( file ).open();
-        } catch( Exception e ) {
-            throw new IllegalArgumentException( e.getMessage() );
+    private void write( String policy, String request, String response ) {
```
Comment (hexec) [inline PolicyDecisionPoint/src/main/java/it/cnr/iit/ucs/pdp/PolicyDecisionPoint.java]: we could put a variadic write function in the JournalInterface

File: UCSLibraries/src/main/java/it/cnr/iit/ucs/journaling/JournalingFileSystem.java [hunk 11]
```diff
@@ -1,29 +1,28 @@
...
+import it.cnr.iit.ucs.properties.base.JournalProperties;
 import it.cnr.iit.utility.FileUtility;
 import it.cnr.iit.utility.errorhandling.Reject;
 
 import journal.io.api.Journal;
 import journal.io.api.Journal.WriteType;
 import journal.io.api.JournalBuilder;
 
-public class PIPJournalHelper {
+public class JournalingFileSystem implements JournalingInterface {
```
Comment (hexec) [inline UCSLibraries/src/main/java/it/cnr/iit/ucs/journaling/JournalingFileSystem.java]: Maybe FileJournal is a better name?

File: UCSLibraries/src/main/java/it/cnr/iit/ucs/journaling/JournalingSyslog.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package it.cnr.iit.ucs.journaling;
+
+import it.cnr.iit.ucs.properties.base.JournalProperties;
+
+public class JournalingSyslog implements JournalingInterface {
```
Comment (hexec) [inline UCSLibraries/src/main/java/it/cnr/iit/ucs/journaling/JournalingSyslog.java]: SyslogJournal or SyslogJournalWrapper

File: PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java (hunk 6 above)
Comment (antoniolm23) [reply PIPReader/src/main/java/it/cnr/iit/ucs/pipreader/PIPReader.java]: done that, here things are different because there is the formatting of the string
